test/replication-luatest/gh_5295_split_brain_test.lua

```diff
@@ -0,0 +1,247 @@
+    end)
+end)
+
+g.before_each(function(cg)
+    -- Check that the servers start synced and with s[1] being leader.
+    -- It's a prerequisite for each test.
+    cg.s[1]:exec(function()
+        require('luatest').assert_equals(box.info.synchro.queue.owner,
+                                         box.info.id, 'First node is leader')
+    end)
+    t.helpers.retrying({}, function()
+        cg.s[num_instances]:assert_follows_upstream(cg.main_id)
+    end)
+end)
+
+-- Drop the partitioned server after each case of split-brain.
+g.after_each(function(cg)
+    cg.s[num_instances]:stop()
+    cg.s[num_instances]:cleanup()
+    cg.cluster.servers[num_instances] = nil
+    num_instances = num_instances - 1
+end)
+
+g.after_all(function(cg)
+    cg.cluster:drop()
+    cg.cluster.servers = nil
+end)
+
+local function update_replication(...)
+    return box.cfg{replication = {...}}
+end
+
+local function partition_nth_instance(cg, n)
+    -- Each partitioning starts on synced servers.
+    cg.s[n]:wait_vclock_of(cg.s[1])
+    cg.s[1]:wait_vclock_of(cg.s[n])
+    cg.s[n]:exec(update_replication, {})
+    cg.box_cfg.replication = repl_list(n - 1)
+    for i = 1, n - 1 do
+        cg.s[i]:exec(update_replication, cg.box_cfg.replication)
+    end
+    return cg.s[n]
+end
+
+local function reconnect_and_check_split_brain(srv, main_id)
+    srv:exec(update_replication, {server.build_instance_uri('s1')})
+    t.helpers.retrying({}, srv.exec, srv, function(main_id)
+        local upstream = box.info.replication[main_id].upstream
+        local t = require('luatest')
+        t.assert_equals(upstream.status, 'stopped', 'replication is stopped')
+        t.assert_str_contains(upstream.message, 'Split-Brain discovered: ',
+                              false, 'split-brain is discovered')
+    end, {main_id})
+end
+
+local function write_promote()
+    local t = require('luatest')
+    t.assert_not_equals(box.info.synchro.queue.owner,  box.info.id,
+                        "Promoting a follower")
+    box.ctl.promote()
+    box.ctl.wait_rw()
+    t.helpers.retrying({}, function()
+        t.assert_equals(box.info.synchro.queue.owner, box.info.id,
+                        "Promote succeeded")
+    end)
+end
+
+local function write_demote()
+    local t = require('luatest')
+    t.assert_equals(box.info.synchro.queue.owner, box.info.id,
+                    "Demoting the leader")
+    box.cfg{election_mode = 'off'}
+    box.ctl.demote()
+    box.cfg{election_mode = 'manual'}
+    t.assert_equals(box.info.synchro.queue.owner, 0, "Demote succeeded")
+end
+
+num_instances = num_instances + 1
+-- Any async transaction performed in an obsolete term means a split-brain.
+g.test_async_old_term = function(cg)
+    local test_server = partition_nth_instance(cg, num_instances)
+    test_server:exec(write_promote)
+    cg.s[1]:exec(function(n) box.space.async:replace{n} end, {num_instances})
+    reconnect_and_check_split_brain(test_server, cg.main_id)
```

Comment:
I don't think I understand. In the tests you separate the last instance s[n]. And promote it. So it gets term > the term of s[1]. Why does it even fail? Shouldn't s[1] just receive latest updates from s[n] and agree that the latter is a leader? And if it should fail somewhy, then why after reconnect s[1] stays fine? Shouldn't it also fail? Or it does fail its downstream to s[b] but it is not checked in the tests?

Reply:
I only connect s[n] to s[1]. Not vice versa. So s[1] remains intact, not knowing that anything bad has happened. All the split brain conditions are checked on s[n].
In this particular case (with async transaction), s[1] would probably follow s[n] just fine, indeed.

Do we need to do something about it? Can we do anything about it? All the split brain detection relies on limbo terms and confirmed_lsn. And neither of these are violated if you connect s[1] to s[n].

Now you might ask, why an async transaction in a old term means a split-brain at all. We came up with this while discussing fencing: https://github.com/tarantool/tarantool/issues/7110#issuecomment-1145868767

The idea is that once the node is not a leader, it shouldn't be able to spoil everyone else's life anyhow.
It wasn't true until we disallow async transactions in an old term.

Reply:
Thanks for the explanation, I see now.

> Do we need to do something about it? Can we do anything about it? All the split brain detection relies on limbo terms and confirmed_lsn. And neither of these are violated if you connect s[1] to s[n].

If you want. I am fine with how it is now too.